### PR TITLE
Revisit the architecture of CompilerCache

### LINKED ISSUE #5912: Revisit the architecture of CompilerCache
@rynowak and I had a look at the CompilerCache and the design of the type looks like it could be refactored \ simplified. Creating a work item for us to revisit this.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #5912

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs [hunk 1]
```diff
@@ -1,133 +1,133 @@
...
+            _compiler = compiler;
+            _compilationCallback = optionsAccessor.Value.CompilationCallback;
+            _logger = logger.CreateLogger<RazorCompiler>();
+
+            _getCompilationResult = GetCompilationResult;
         }
 
         private ICompilerCache CompilerCache => _compilerCacheProvider.Cache;
 
         public CompilerCacheResult Compile(string relativePath)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorCompiler.cs]: The goal was to make this one method the user needs to invoke to get caching + compilation. `DefaultRoslynCompilation` required independently caching the result which isn't very obvious.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+using System.Linq;
+using Microsoft.AspNetCore.Diagnostics;
+using Microsoft.AspNetCore.Mvc.Razor.Compilation;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    internal static class CompilationFailedResultFactory
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: There's a bit of overlap in terms of creating CompilationFailedResult from both Razor and Roslyn. Combining the two here.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ICompilationService.cs [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 
+using System;
 using Microsoft.AspNetCore.Razor.Language;
 
 namespace Microsoft.AspNetCore.Mvc.Razor.Compilation
 {
     /// <summary>
     /// Provides methods for compilation of a Razor page.
     /// </summary>
+    [Obsolete("This type will be removed in a future version. Use RazorCompiler instead.")]
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ICompilationService.cs]: I'll check with @Eilon if we can remove the type entirely or mark it as obsolete for now and remove it in a later version.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ICompilationService.cs]: If it's not used for anything we should just nuke it IMO.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        // error CS0234: The type or namespace name 'C' does not exist in the namespace 'N' (are you missing
+        // an assembly reference?)
+        private const string CS0234 = nameof(CS0234);
+        // error CS0246: The type or namespace name 'T' could not be found (are you missing a using directive
+        // or an assembly reference?)
+        private const string CS0246 = nameof(CS0246);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: Why do this at all? Why not just have a string constant?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+        private static string GetFilePath(RazorCodeDocument codeDocument, Diagnostic diagnostic)
+        {
+            if (diagnostic.Location == Location.None)
+            {
+                return codeDocument.Source.FileName;
+            }
+
+            return diagnostic.Location.GetMappedLineSpan().Path;
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: Most of this code seems generic enough that it should live in `Microsoft.CodeAnalysis.Razor`

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs [hunk 6]
```diff
@@ -67,27 +79,26 @@ public CompilerCache(IFileProvider fileProvider)
         /// <inheritdoc />
         public CompilerCacheResult GetOrAdd(
             string relativePath,
-            Func<string, CompilerCacheContext> cacheContextFactory)
+            Func<string, CompilationResult> compile)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: Remind me again why this needs to be passed in here?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs [hunk 7]
```diff
@@ -126,19 +136,19 @@ public CompilerCache(IFileProvider fileProvider)
 
                 cacheEntryOptions = new MemoryCacheEntryOptions();
 
-                compilerCacheContext = cacheContextFactory(normalizedPath);
-                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(compilerCacheContext.ProjectItem.Path));
-                if (!compilerCacheContext.ProjectItem.Exists)
+                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(normalizedPath));
+                var projectItem = _templateEngine.Project.GetItem(normalizedPath);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: I don't feel like this should be coupled to project items and the template engine in particular. The template engine is the server, what's the abstraction for a client?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs [hunk 8]
```diff
@@ -67,27 +79,26 @@ public CompilerCache(IFileProvider fileProvider)
         /// <inheritdoc />
         public CompilerCacheResult GetOrAdd(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: Would be better to return `Task<>` here and let the caller decide if they should block or not?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: Weird design - where the compiler cache did not know about the compiler. It might be cleaner to change this so that cache just calls in to the compiler from here.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: If this ends up being something like:
```
public interface IViewCompiler
{
    public Task<CompiledViewDescriptor> CompileAsync(string relativePath)
}
```

It's way simpler, and the fact that we have compilation at all is just a detail.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ICompilationService.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ICompilationService.cs]: FYI @Eilon since the interface was in a public namespace.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: Because it's awesome?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs (hunk 7 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: We need some API to give us the import hierarchy \ import file name. Nothing outside of the template engine has those bits of data today.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs]: I'm not pushing back on the use of the template engine, I'm just suggesting maybe it's not a **direct** relationship

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    internal static class CompilationFailedResultFactory
+    {
+        // error CS0234: The type or namespace name 'C' does not exist in the namespace 'N' (are you missing
+        // an assembly reference?)
+        private const string CS0234 = nameof(CS0234);
+        // error CS0246: The type or namespace name 'T' could not be found (are you missing a using directive
+        // or an assembly reference?)
+        private const string CS0246 = nameof(CS0246);
+
+        public static CompilationResult Create(
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: AllI see is allocations

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: ?

**Comment by pranavkm** [comment]
Still in design. 🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs (hunk 9 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs]: LINQ

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Compilation
+{
+    public struct CompiledViewDescriptor
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs]: There's really no reason to make this a struct since they get cached. And if you are going to make a struct, immutable, always.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/IViewCompilerProvider.cs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Compilation
+{
+    public interface IViewCompilerProvider
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/IViewCompilerProvider.cs]: Why is this needed?

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        public string RelativePath { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="RazorViewAttribute"/> decorating the sview.
+        /// </summary>
+        public RazorViewAttribute ViewAttribute { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs]: Don't expose the attribute directly here, put the type and path on this type.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs [hunk 13]
```diff
@@ -40,7 +40,7 @@ public void PopulateFeature(IEnumerable<ApplicationPart> parts, ViewsFeature fea
 
                 foreach (var item in viewContainer.ViewInfos)
                 {
-                    feature.Views[item.Path] = item.Type;
+                    feature.ViewMetadata.Add(new RazorViewAttribute(item.Path, item.Type));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs]: Would it be possible to just create the view descriptors here?

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs]: Ah I see what you're doing. Consider just making this class abstract `CompiledItemDescriptor` and having two concrete subclasses (view + page)

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/IViewCompilerProvider.cs (hunk 11 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/IViewCompilerProvider.cs]: I was following the same pattern as the `ICompilerCacheProvider`. Instantiating the cache requires reading precompiled view metadata and we wanted to avoid doing that at startup (because it involves a bunch of disk I/O looking for the precompiled assembly). The alternative was to lazy initialize things inside the `RazorViewCompiler` which ends up being messier.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/IViewCompilerProvider.cs]: Yeah I think this is fine. The other option is to just use a factory method with DI.

The problem with either approach is that we need to be careful to lazy-initialize this **everywhere** or else it doesn't add value. I think what you've done here is fine.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationParts/CompiledPageFeature.cs [hunk 14]
```diff
@@ -2,11 +2,12 @@
...
 
 using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure;
 
 namespace Microsoft.AspNetCore.Mvc.ApplicationParts
 {
     public class CompiledPageInfoFeature
     {
-        public IList<CompiledPageInfo> CompiledPages { get; } = new List<CompiledPageInfo>();
+        public IList<RazorPageAttribute> CompiledPages { get; } = new List<RazorPageAttribute>();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationParts/CompiledPageFeature.cs]: Why isn't this `CompiledViewDescriptor`?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/RazorPageAttribute.cs [hunk 15]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Mvc.Razor.Compilation;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    public class RazorPageAttribute : RazorViewAttribute
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/RazorPageAttribute.cs]: I'm not a fan of this, but I think we need to get the rest of this change done.
